client-runtime/src/main/java/com/microsoft/rest/v2/http/NettyClient.java

```diff
@@ -91,11 +129,41 @@ public void channelCreated(Channel ch) throws Exception {
             }, eventLoopGroup.executorCount() * 2);
         }
 
+        @SuppressWarnings("unchecked")
         private NettyAdapter(int eventLoopGroupSize, int channelPoolSize) {
-            this.eventLoopGroup = new NioEventLoopGroup(eventLoopGroupSize);
+            MultithreadEventLoopGroup eventLoopGroup;
+            Class<? extends SocketChannel> channelClass;
+            try {
+                final String osName = System.getProperty("os.name");
+                if (osName.contains("Linux")) {
+                    eventLoopGroup = (MultithreadEventLoopGroup) Class.forName("io.netty.channel.epoll.EpollEventLoopGroup").getConstructor(Integer.TYPE).newInstance(eventLoopGroupSize);
+                    channelClass = (Class<? extends SocketChannel>) Class.forName("io.netty.channel.epoll.EpollSocketChannel");
+                } else if (osName.contains("Mac")) {
+                    eventLoopGroup = (MultithreadEventLoopGroup) Class.forName("io.netty.channel.kqueue.KQueueEventLoopGroup").getConstructor(Integer.TYPE).newInstance(eventLoopGroupSize);
+                    channelClass = (Class<? extends SocketChannel>) Class.forName("io.netty.channel.kqueue.KQueueSocketChannel");
+                } else {
+                    eventLoopGroup = new NioEventLoopGroup();
+                    channelClass = NioSocketChannel.class;
+                }
+            } catch (Exception e) {
+                String message = e.getMessage();
+                if (message == null) {
+                    Throwable cause = e.getCause();
+                    if (cause != null) {
+                        message = cause.getMessage();
+                    }
+                }
+
+                LoggerFactory.getLogger(NettyAdapter.class).debug("Exception when obtaining native EventLoopGroup and SocketChannel: " + message);
+                eventLoopGroup = new NioEventLoopGroup();
+                channelClass = NioSocketChannel.class;
+            }
+            this.eventLoopGroup = eventLoopGroup;
```

Comment:
This is identical to the code on line 87. Can you combine these somehow?

Reply:
I'd like to. The main struggle is the reference to `eventLoopGroup.executorCount()`, which makes it hard to simply delegate the parameterless constructor to a constructor which explicitly takes a pool size and event loop group size.